From 16c6bb669d013e8a6cc8ead708cf68509469b943 Mon Sep 17 00:00:00 2001
From: Lukas Zeller <luz@plan44.ch>
Date: Fri, 11 Dec 2020 22:39:56 +0100
Subject: [PATCH] vfs.c: set_global_fd_sets() was broken because it iterated
 trough global fds for each vfs but did not check vfs index when translating
 to local_fd

---
 components/vfs/vfs.c | 32 ++++++++++++++++++--------------
 1 file changed, 18 insertions(+), 14 deletions(-)

diff --git a/components/vfs/vfs.c b/components/vfs/vfs.c
index 005d050d4..9bcd52f39 100644
--- a/components/vfs/vfs.c
+++ b/components/vfs/vfs.c
@@ -804,20 +804,24 @@ static int set_global_fd_sets(const fds_triple_t *vfs_fds_triple, int size, fd_s
         if (item->isset) {
             for (int fd = 0; fd < MAX_FDS; ++fd) {
                 const int local_fd = s_fd_table[fd].local_fd; // single read -> no locking is required
-                if (readfds && esp_vfs_safe_fd_isset(local_fd, &item->readfds)) {
-                    ESP_LOGD(TAG, "FD %d in readfds was set from VFS ID %d", fd, i);
-                    FD_SET(fd, readfds);
-                    ++ret;
-                }
-                if (writefds && esp_vfs_safe_fd_isset(local_fd, &item->writefds)) {
-                    ESP_LOGD(TAG, "FD %d in writefds was set from VFS ID %d", fd, i);
-                    FD_SET(fd, writefds);
-                    ++ret;
-                }
-                if (errorfds && esp_vfs_safe_fd_isset(local_fd, &item->errorfds)) {
-                    ESP_LOGD(TAG, "FD %d in errorfds was set from VFS ID %d", fd, i);
-                    FD_SET(fd, errorfds);
-                    ++ret;
+                if (local_fd>=0 && s_fd_table[fd].vfs_index==i) {
+                    ESP_LOGD(TAG, "***  fd=%d, VFS ID %d, vfs_index[fd]=%d, local_fd=%d,", fd, i, s_fd_table[fd].vfs_index, local_fd);
+                    ESP_LOGD(TAG, "    item->readfds=%lx, item->errorfds=%lx", item->readfds.fds_bits[0], item->errorfds.fds_bits[0]);
+                    if (readfds && esp_vfs_safe_fd_isset(local_fd, &item->readfds)) {
+                        ESP_LOGD(TAG, "FD %d in readfds was set from VFS ID %d", fd, i);
+                        FD_SET(fd, readfds);
+                        ++ret;
+                    }
+                    if (writefds && esp_vfs_safe_fd_isset(local_fd, &item->writefds)) {
+                        ESP_LOGD(TAG, "FD %d in writefds was set from VFS ID %d", fd, i);
+                        FD_SET(fd, writefds);
+                        ++ret;
+                    }
+                    if (errorfds && esp_vfs_safe_fd_isset(local_fd, &item->errorfds)) {
+                        ESP_LOGD(TAG, "FD %d in errorfds was set from VFS ID %d", fd, i);
+                        FD_SET(fd, errorfds);
+                        ++ret;
+                    }
                 }
             }
         }
-- 
2.27.0

